### PR TITLE
feat: implement oauth 2.0 for azdo with bot framework sdk

### PR DESCRIPTION
**Pull request by hund030**

### CODE REVIEW

**Review by Siglud — APPROVED** [review]

File: bot/codeSharingBot.ts [hunk 1]
```diff
@@ -20,10 +22,37 @@ export class CodeSharingBot extends TeamsActivityHandler {
     // Unfurling link contains `github`.
     if (url.includes("github.com")) {
       return await handleGitHubUrl(url);
+    } else if (url.includes(".visualstudio.com")) {
+      const adapter = context.adapter as CloudAdapter;
+      const userTokenClient = context.turnState.get(adapter.UserTokenClientKey);
```
Comment (yukun-dong) [inline bot/codeSharingBot.ts]: How about using  `get<UserTokenClient>` to make the type for `userTokenClient` explicit.

**Review by yukun-dong — APPROVED** [review]

File: templates/azure/teamsFx/webapp.bicep [hunk 2]
```diff
@@ -7,19 +7,16 @@ param provisionOutputs object
...
   name: '${botWebAppName}/appsettings'
   properties: union({
-    BOT_ID: botAadAppClientId // ID of your bot
-    BOT_PASSWORD: botAadAppClientSecret // Secret of your bot
-    IDENTITY_ID: provisionOutputs.identityOutput.value.identityClientId // User assigned identity id, the identity is used to access other Azure resources
-  }, currentAppSettings)
+      MicrosoftAppId: botAadAppClientId // ID of your bot
+      MicrosoftAppPassword: botAadAppClientSecret // Secret of your bot
+      MicrosoftAppType: 'MultiTenant'
+      ConnectionName: 'azdo2'
```
Comment (yukun-dong) [inline templates/azure/teamsFx/webapp.bicep]: Is this supposed to be "azdo"?

File: README.md [hunk 3]
```diff
@@ -1,18 +1,73 @@
...
 # Note
+
 - API for code rendering: see in function `renderCodeWithAPI` in `reqHelper.ts`, modify to another API if needed.
 - Link unfurling: see in `templates/appPackage/manifest.template.json`, modify `github.com` to other domains if needed.
 
 # Features
+
 ## Link Unfurling
+
 - GitHub permalink: It is like `https://github.com/OfficeDev/TeamsFx-Samples/blob/master/test.py#L1-L6?x=a&y=b`.
```
Comment (eriolchan) [inline README.md:68]: Since the OAuth is for AzDO repository, do we need to add AzDO permalink as well?

File: bot/codeSharingBot.ts [hunk 4]
```diff
@@ -50,10 +57,15 @@ async function createCardCommand(
   // If URL contains `github`, use GitHub API route.
   if (url.includes("github.com")) {
     return await handleGitHubUrl(url);
+  } else if (url.includes(".visualstudio.com")) {
```
Comment (eriolchan) [inline bot/codeSharingBot.ts:69]: It looks the same code logic in handleTeamsAppBasedLinkQuery(), and it is better to refactor it later for long-term code maintenance.

File: bot/helper/reqHelper.ts [hunk 5]
```diff
@@ -169,9 +170,10 @@ async function reqInfoFromAzDOAPI(
   projectName: string,
   repoName: string,
   path: string,
-  ref: string
+  ref: string,
+  accessToken?: string
 ) {
-  const token = getAzDOToken();
+  const token = accessToken ? `Bearer ${accessToken}` : getAzDOToken();
```
Comment (eriolchan) [inline bot/helper/reqHelper.ts]: What's the difference between "access token" and getAzDOToken()?

File: templates/azure/teamsFx/webapp.bicep [hunk 6]
```diff
@@ -7,19 +7,16 @@ param provisionOutputs object
...
+      IDENTITY_ID: provisionOutputs.identityOutput.value.identityClientId // User assigned identity id, the identity is used to access other Azure resources
+    }, currentAppSettings)
```
Comment (eriolchan) [inline templates/azure/teamsFx/webapp.bicep:21]: I have a question on "As we adding more feature to this Teams App, do we need to add some unit test cases"?

File: bot/codeSharingBot.ts (hunk 1 above)
Comment (hund030) [reply bot/codeSharingBot.ts]: Resolved. Thanks~

File: bot/codeSharingBot.ts (hunk 4 above)
Comment (hund030) [reply bot/codeSharingBot.ts:69]: Refactored. Thanks~

File: bot/helper/reqHelper.ts (hunk 5 above)
Comment (hund030) [reply bot/helper/reqHelper.ts]: `getAzDOToken` retrieves token string from environment variables. Removed.

File: README.md (hunk 3 above)
Comment (hund030) [reply README.md:68]: Added.

File: templates/azure/teamsFx/webapp.bicep (hunk 6 above)
Comment (hund030) [reply templates/azure/teamsFx/webapp.bicep:21]: Yes. I think we need to setup tests for the app. And I think this should be a sample for setting up unit test for Teams App.

**Review by JerryYangKai — APPROVED** [review]

**Pull request merged**